### PR TITLE
fastem: update beamshift correction to use estimate_grid_orientation_…

### PR DESCRIPTION
**Pull request by tepals**
…from_img instead of FindGridSpots

### CODE REVIEW

File: src/odemis/acq/fastem.py [hunk 1]
```diff
@@ -783,31 +785,33 @@ def correct_beam_shift(self):
...
         centered on the mppc detector.
         """
+        pixel_size = self._ccd.pixelSize.value
+        magnification = self._lens.magnification.value
+        refractive_index = self._lens.refractiveIndex.value
+        numerical_aperture = self._lens.numericalAperture.value
+        wavelength = 550e-9
+        sigma = (magnification / pixel_size) * synthetic.psf_sigma_wffm(
+            refractive_index, numerical_aperture, wavelength
+        )
```
Comment (tepals) [inline src/odemis/acq/fastem.py:795]: @effting is this the correct way of calculating sigma?

Comment (effting) [reply src/odemis/acq/fastem.py:795]: It should be better to use `ccd.pointSpreadFunctionSize` from `src/odemis/model/_components.py`

Comment (effting) [reply src/odemis/acq/fastem.py:795]: Could also be `self._ccd.spotSize`. Need to check which one to use

File: src/odemis/acq/fastem.py [hunk 2]
```diff
@@ -783,31 +785,33 @@ def correct_beam_shift(self):
...
         # asap=False: wait until new image is acquired (don't read from buffer)
         ccd_image = self._ccd.data.get(asap=False)
-
-        # Find the location of the spots on the diagnostic camera.
-        spot_coordinates, *_ = FindGridSpots(ccd_image, (8, 8))
-
-        # Transform the spots from the diagnostic camera coordinate system to a right-handed coordinate
-        # system with the origin in the bottom left.
-        spot_coordinates[:, 1] = ccd_image.shape[1] - spot_coordinates[:, 1]  # [px]
+        tform = estimate_grid_orientation_from_img(ccd_image, (8, 8), AffineTransform, sigma)
```
Comment (tepals) [inline src/odemis/acq/fastem.py:798]: @effting we were using an `AffineTransform` in `FindGridSpots`, I see that in your code you use a `SimilarityTransform`. Did you try both for the tests with missing spots, and was one performing "better"?

File: src/odemis/acq/fastem.py (hunk 1 above)
Comment (effting) [reply src/odemis/acq/fastem.py:795]: Indeed, should be `self._ccd.pointSpreadFunctionsize.value`

File: src/odemis/acq/fastem.py (hunk 2 above)
Comment (effting) [reply src/odemis/acq/fastem.py:798]: The main goal of the beamshift correction is to control (limit) cross-talk. We need to project the image of the multi-probe pattern on the MPPC sensor array. We can only control the scale, rotation, and translation so it makes sense to me to not fit the anisotropic scaling and shear here.

File: src/odemis/acq/fastem.py [hunk 3]
```diff
@@ -506,6 +508,8 @@ def __init__(self, scanner, multibeam, descanner, detector, stage, ccd, beamshif
         self._pre_calibrations = pre_calibrations
         self._settings_obs = settings_obs
 
+        self.path = fastem_util.create_image_dir("beam_shift_cor")
```
Comment (nandishjpatel) [inline src/odemis/acq/fastem.py:511]:
```suggestion
        self.path = fastem_util.create_image_dir("beam-shift-correction")
```

**Review by nandishjpatel — APPROVED** [review]

File: src/odemis/acq/fastem.py [hunk 4]
```diff
@@ -783,34 +787,49 @@ def correct_beam_shift(self):
...
+        fav_pos_active = self._ccd.getMetadata()[model.MD_FAV_POS_ACTIVE]
+        # If i and j are not in the metadata, use x and y instead.
+        try:
+            i = fav_pos_active["i"]
+        except KeyError:
+            i = fav_pos_active["x"]
+        try:
+            j = fav_pos_active["j"]
+        except KeyError:
+            j = ccd_image.shape[1] - fav_pos_active["y"]
```
Comment (pieleric) [inline src/odemis/acq/fastem.py:810]: Why wouldn't it be there, while x/y would be there? Old version of the calibration function? If so, please add this info in the comment, and indicate the x/y code can be dropped soon.

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/acq/fastem.py [hunk 5]
```diff
@@ -783,34 +787,49 @@ def correct_beam_shift(self):
...
+        #  on the diagnostic camera. Therefore we need to invert the shift.
+        #  When the beamshift component is fixed, this should be removed.
+        shift_m *= -1
+        beam_shift_cor = shift_m / magnification  # [m]
         # Convert the shift from pixels to meters
-        pixel_size = self._ccd.pixelSize.value
-        magnification = self._lens.magnification.value
-        shift_m = shift * pixel_size / magnification  # [m] pixel size diagnostic camera divided by 40x magnification
-        logging.debug("Beam shift adjustment required due to stage magnetic field: {} [m]".format(shift_m))
+        logging.debug("Beam shift adjustment required due to stage magnetic field: {} [m]".format(beam_shift_cor))
```
Comment (effting) [inline src/odemis/acq/fastem.py:822]: You can drop the "due to stage magnetic field"

**Review by effting — APPROVED** [review]

**Review by nandishjpatel — APPROVED** [review]

**Pull request merged**